test/actor_tests.cpp

```diff
@@ -0,0 +1,222 @@
+}
+
+/**************************************************************************************************/
+
+BOOST_AUTO_TEST_CASE(actor_regularity_void) {
+    std::size_t count{0};
+    stlab::actor<void> empty_ctor;
+
+    stlab::actor<void> default_ctor(stlab::default_executor, "foo"); // default construction
+    stlab::await(default_ctor.send([&] { ++count; }));
+    BOOST_REQUIRE(count == 1);
+
+    stlab::actor<void> copy_ctor(default_ctor); // copy construction
+    stlab::await(copy_ctor.send([&] { ++count; }));
+    BOOST_REQUIRE(count == 2);
+
+    stlab::actor<void> move_ctor(std::move(default_ctor)); // move construction
+    stlab::await(move_ctor.send([&] { ++count; }));
+    BOOST_REQUIRE(count == 3);
+
+    stlab::actor<void> copy_assign = move_ctor; // copy assignment
+    stlab::await(copy_assign.send([&] { ++count; }));
+    BOOST_REQUIRE(count == 4);
+
+    stlab::actor<void> move_assign = std::move(move_ctor); // move assignment
+    stlab::await(move_assign.send([&] { ++count; }));
+    BOOST_REQUIRE(count == 5);
+
+    // equality comparable
+    stlab::actor<void> a(stlab::default_executor, "a");
+    stlab::actor<void> b(stlab::default_executor, "b");
+    BOOST_REQUIRE(a != b);    // tests operator!=
+    BOOST_REQUIRE(!(a == b)); // tests operator==
+
+    BOOST_REQUIRE(a == a); // tests operator==
+}
+
+/**************************************************************************************************/
+
+BOOST_AUTO_TEST_CASE(actor_send_to_void) {
+    {
+        stlab::actor<int> a(stlab::default_executor, "send_getting_void");
+        stlab::future<void> f = a.send(increment);
+
+        stlab::await(f);
+
+        BOOST_REQUIRE(f.get_try());
+    }
+
+    {
+        stlab::actor<void> a(stlab::default_executor, "send_getting_void");
+        std::atomic_bool sent{false};
+        stlab::future<void> f = a.send([&] { sent = true; });
+
+        stlab::await(f);
+
+        BOOST_REQUIRE(sent);
+    }
+}
+
+/**************************************************************************************************/
+
+BOOST_AUTO_TEST_CASE(actor_send_to_value) {
+    {
+        stlab::actor<int> a(stlab::default_executor, "send_getting_value", 42);
+        stlab::future<int> f = a.send([](auto& x) { return x; });
+        int result = stlab::await(f);
+
+        BOOST_REQUIRE(result == 42);
+    }
+
+    {
+        stlab::actor<void> a(stlab::default_executor, "send_getting_value");
+        stlab::future<int> f = a.send([]() { return 42; });
+        int result = stlab::await(f);
+
+        BOOST_REQUIRE(result == 42);
+    }
+}
+
+/**************************************************************************************************/
+
+BOOST_AUTO_TEST_CASE(actor_then_from_void) {
+    {
+        stlab::actor<int> a(stlab::default_executor, "send_then_from_void");
+        stlab::future<void> f0 = a.send(increment_by, 42);
+        stlab::future<int> f1 = a.then(stlab::future<void>(f0), [](auto& x) { return x; });
```

Comment:
Why x is ref?

Reply:
Fixed.

Reply:
These first parameters are the `actor`'s task local state - in this case an `int`. The lambda has exclusive access to this task state and is allowed to modify it if it so chooses. I have an example where this is done. In this case there isn't much difference between an `int` and an `int&`, but the `actor`'s task local state can be anything, and in such case the pattern to take a reference will save unnecessary copies.

Reply:
I have made the non-ref change here, but hopefully the above comment adds clarity as to why a reference here is both allowed and desired.